flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartitionReadSchedulerTest.java

```diff
@@ -192,6 +203,132 @@ public void testOnReadBufferRequestError() throws Exception {
         assertAllResourcesReleased();
     }
 
+    @Test
+    public void testBufferRequestTimeoutWithoutRecycle() throws Exception {
+        occupyBuffers();
+
+        SortMergeSubpartitionReader subpartitionReader =
+                readScheduler.createSubpartitionReader(
+                        new NoOpBufferAvailablityListener(), 0, partitionedFile);
+
+        Set<SortMergeSubpartitionReader> allReaders = new HashSet<>();
+        allReaders.add(subpartitionReader);
+        readScheduler.allocateBuffers(new PriorityQueue<>(allReaders));
+
+        assertExpectedTimeoutException(subpartitionReader.getFailureCause());
+    }
+
+    @Test
+    public void testBufferRequestSuccessWithRecycle() throws Exception {
+        List<MemorySegment> buffers = occupyBuffers();
+
+        // Recycle the buffer after some time
+        Random random = new Random();
+        int recycleBufferWaitMs = random.nextInt(2999);
+        startBufferRecycleThread(bufferPool, buffers, recycleBufferWaitMs, false);
+
+        SortMergeSubpartitionReader subpartitionReader =
+                readScheduler.createSubpartitionReader(
+                        new NoOpBufferAvailablityListener(), 0, partitionedFile);
+
+        Set<SortMergeSubpartitionReader> allReaders = new HashSet<>();
+        allReaders.add(subpartitionReader);
+        readScheduler.allocateBuffers(new PriorityQueue<>(allReaders));
+
+        assertNull(subpartitionReader.getFailureCause());
+    }
+
+    @Test
+    public void testBufferRequestSuccessAfterRecycle() throws Exception {
+        List<MemorySegment> buffers = occupyBuffers();
+
+        SortMergeSubpartitionReader subpartitionReader =
+                readScheduler.createSubpartitionReader(
+                        new NoOpBufferAvailablityListener(), 0, partitionedFile);
+        Set<SortMergeSubpartitionReader> allReaders = new HashSet<>();
+        allReaders.add(subpartitionReader);
+        // Recycle the buffer after 1s
+        startBufferRecycleThread(bufferPool, buffers, 1000, false);
+
+        Thread.sleep(3500);
```

Comment:
The request timeout is set to 3s in `testRequestTimeoutIsRefreshedAndSuccess`. Because the inner sleep time in the method `BatchShuffleReadBufferPool#requestBuffers` is 2s, I think the timeout time should be greater than that value. 

After the request timeout is set to 3s, the `firstRecycleTimeMs` is set to 2s and the `secondRecycleTimeMs` is set to 4s to prove the timeout time is refreshed and increased when buffers are recycled.